### DISCUSSION TITLE
jbig2 support

### DISCUSSION THREAD

**Discussion by jknockaert** [ImageMagick/ImageMagick#3182] [Help]
When I run `identify -list compress` it shows JBIG2 support. However, JBIG2 compression seems not to be supported for pdf image files: https://github.com/ImageMagick/ImageMagick/issues/623. So I was wondering what JBIG2 compression support there actually is in imagemagick. Is it the lossless or the lossy version of JBIG2?

**Comment by urban-warrior** [comment]
We support the JBIG image format which sets the image compression value to JBIG2 when reading a JBIG image.

Post one or two PDF's that are already JBIG compressed.  We'll see if we can add a patch to the PDF reader to read JBIG-encoded PDF image files.

**Reply by jknockaert** [reply]
I was rather thinking of converting a tiff into a JBIG2 compressed PDF.

**Comment by urban-warrior** [comment]
We're willing to investigate supporting reading/writing JBIG2 compessed PDF, however, we first need a couple of sample PDF images that are already JBIG2-compressed.  Once you post a link here to a couple of PDF image files we can use as a reference, we'll begin the work to support JBIG2.

**Reply by AmandaUCSC** [reply]
ETA: would this be the correct solution to my issue, it seems really out of date and only for jbig1 https://www.cl.cam.ac.uk/~mgk25/jbigkit/

Hi, I'm having this issue... I can't post the content of one of our PDFs, but I created a mock PDF (see attached) using the same format as the others (jbig2). Here's the context/issue: 
- Imagemagick (version 6.9.10) is installed, for use with Omeka and the PDF to text plug-in, which allows the uploading of the PDF and then Imagemagick/Ghostscript (v 9.25) to create the thumbnails.
- The PDF can be viewed without a problem, the thumbnails show up only as a white/blank image. There's no error thrown, just the white background image.
- With one of the Omeka developers, I've traced the issue back to jbig2, which I believe is the problem because older PDFs without jbig2 compression seem to be fine; the plug-in still creates those thumbnails with no issue.
- The PDFs were/are created in ABBYY FineReader with the original scans as JPGs or TIFFs, then "saved" (exported) as searchable PDFs. So I guess it's ABBYY that's compressing them as jbig2 files, but that's what I'm currently trying to trace back... EDIT: yes it's that. ABBYY by default converts all b and w images in PDFs to "jbig2 (lossless)"
- This mock file was created using the same scanner + ABBYY Reader process: 
[MA Thesis Registration Forms.pdf](https://github.com/ImageMagick/ImageMagick/files/10131995/MA.Thesis.Registration.Forms.pdf)

**Reply by fmw42** [reply]
I do not know where JBIG2 comes in, but identify show an unknown compression.  Nevertheless with IM 6.9.12-67 Q16 on Mac OSX Monterey and GS 9.56, I can convert your pdf to jpg just fine

`convert MA.Thesis.Registration.Forms.pdf[0] x.jpg`

![x](https://user-images.githubusercontent.com/13458218/205114679-f0ced198-435a-4e00-9c14-68080bce2d65.jpg)

Please clarify further how this can be tested without all your other tools

**Reply by urban-warrior** [reply]
Recall we need a couple of JBIG2-compressed PDF's so we can use them to validate any JBIG2 support we implement.  The image you posted is not JBIG2-compressed.  Instead there are numerous objects that are Flate (png) and JPX (JPEG 2000) compressed.  ImageMagick currently supports both these compression algorithms when writing PDF images.

**Reply by AmandaUCSC** [reply]
> Recall we need a couple of JBIG2-compressed PDF's so we can use them to validate any JBIG2 support we implement. The image you posted is not JBIG2-compressed. Instead there are numerous objects that are Flate (png) and JPX (JPEG 2000) compressed. ImageMagick currently supports both these compression algorithms when writing PDF images.

Thanks -- did so below.

**Comment by AmandaUCSC** [comment]
ETA: See attached. I split up the images in ABBYY and processed them into three separate PDFs. Apologies for the confusion earlier. I'm also wondering if an upgrade to IM/GS would make a difference here, i.e. IM 7.1 with GS 9.56. Also, just to reiterate, it is specifically a thumbnail creation issue (if that makes a difference).

[TEST JBIG2 3.pdf](https://github.com/ImageMagick/ImageMagick/files/10141815/TEST.JBIG2.3.pdf)
[TEST JBIG2 2.pdf](https://github.com/ImageMagick/ImageMagick/files/10141816/TEST.JBIG2.2.pdf)
[TEST JBIG2.pdf](https://github.com/ImageMagick/ImageMagick/files/10141817/TEST.JBIG2.pdf)

Ah shoot you're right. I'll post 2-3 when I get a chance. I'll use the same
PDF but process it as black and white images in ABBYY then repost later.
Because that is how the other PDFs are processed and are indeed jbig2 then.
I forgot to do that for these, which are color images and not b and w.

**Comment by AmandaUCSC** [comment]
I just wanted to follow up on this and say that I figured out the issue. The server actually hadn't run the update to Ghostscript 9.25; I thought it had but searching through the package updates, that one somehow had errored out. So in fact the server was still running an older version... 9.09 or 9.10 I think? Once I updated to 9.25, ImageMagick created the thumbnail no problem.

**Reply by urban-warrior** [reply]
We're using Ghostscript 9.56 and it cannot grok your sample PDF's.  As we leverage Ghostscript to render PDF's, unless and until Ghostscript can properly render them, ImageMagick will not be able to convert them to other image formats.  However, in the mean-time we can attempt to add JBIG2 support when writing  in the PDF format.
